plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,282 @@
+
+    private static final TimeValue delay = TimeValue.ZERO;
+    /**
+     * The internal OpenSearch thread pool that execute async processing and exporting tasks
+     */
+    private final ThreadPool threadPool;
+
+    /**
+     * enable insight data collection
+     */
+    private final Map<MetricType, Boolean> enableCollect = new HashMap<>();
+
+    private int topNSize = QueryInsightsSettings.DEFAULT_TOP_N_SIZE;
+
+    private TimeValue windowSize = TimeValue.timeValueSeconds(QueryInsightsSettings.DEFAULT_WINDOW_SIZE);
+    /**
+     * The internal thread-safe store that holds the top n queries by latency query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByLatencyStore;
+    /**
+     * The internal thread-safe store that holds the top n queries by cpu query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByCpuStore;
+    /**
+     * The internal thread-safe store that holds the top n queries by jvm query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByJvmStore;
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart = new HashMap<>();
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        this.topQueriesByLatencyStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingLong(a -> a.getMeasurement(MetricType.LATENCY).getValue().longValue())
+        );
+        this.topQueriesByCpuStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.CPU).getValue().doubleValue())
+        );
+        this.topQueriesByJvmStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.JVM).getValue().doubleValue())
+        );
+        this.threadPool = threadPool;
+    }
+
+    private PriorityBlockingQueue<SearchQueryRecord> getStoreByMetricType(MetricType type) {
+        if (type == MetricType.LATENCY) {
+            return topQueriesByLatencyStore;
+        } else if (type == MetricType.CPU) {
+            return topQueriesByCpuStore;
+        } else if (type == MetricType.JVM) {
+            return topQueriesByJvmStore;
+        } else {
+            throw new IllegalArgumentException(
+                "Invalid metricType [" + type + "]" + " should be one of [" + MetricType.allMetricTypes() + "]"
+            );
+        }
+    }
+
+    /**
+     * Ingest the query data into local stores
+     *
+     * @param metricType {@link MetricType}
+     * @param record the record to ingest
+     */
+    public void addRecord(final MetricType metricType, SearchQueryRecord record) {
```

Comment:
On thinking more, I think `addRecord()` should just accept `SearchQueryRecord`. It should parse the record to find all metric types associated with it and then add it to respective priority queues.  The advantage is, if consumer of this service wants to combine multiple metricType into a single record, which is very likely, then they don't have to call it multiple times for each metric type.

Reply:
Let me refactor this part to make the logic simpler